system/Database/BasePreparedQuery.php

```diff
@@ -147,17 +147,18 @@ abstract public function _execute(array $data): bool;
     abstract public function _getResult();
 
     /**
-     * Explicitly closes the statement.
+     * Explicitly closes the prepared statement.
      */
-    public function close()
+    public function close(): bool
     {
-        if (! is_object($this->statement) || ! method_exists($this->statement, 'close')) {
-            return;
-        }
-
-        $this->statement->close();
+        return $this->_close();
     }
 
+    /**
+     * The database dependant version of the close method.
+     */
+    abstract public function _close(): bool;
```

Comment:
Why is this `public`? Shouldn't this be `protected` since you are using the template pattern for the `close` method?

Reply:
Indeed, it is `public`, just like the declarations of prepare/_prepare and execute/_execute methods

But if the database-dependent part should be protected, shouldn't it be for these 3 methods also ?

Reply:
Hmm... that should be wrong. That would be unnecessary visibility for those methods. However, we cannot change the visibilities of `_prepare` and `_execute` as it would be a BC break.

Let us start the correct way with this one and do the others in separate PRs.

Reply:
OK, i've changed the visibility to "protected"